File: apps/api/src/realtime/redis-pubsub.ts
import type Redis from "ioredis";
import type { SSEManager } from "./sse-manager.js";

const ORDER_CHANNEL = "order:status";

export interface OrderStatusEvent {
  orderId: string;
  status: string;
  updatedAt: string;
}

/**
 * Wraps Redis pub/sub to broadcast order status events across API instances.
 * The subscriber receives published messages and forwards them to connected
 * SSE clients via the SSEManager.
 */
export class RedisPubSub {
  constructor(
    private pub: Redis,
    private sub: Redis,
    private sseManager: SSEManager
  ) {}

  /**
   * Subscribe to the order status channel and route incoming messages
   * to the appropriate SSE connections.
   */
  async start(): Promise<void> {
    await this.sub.subscribe(ORDER_CHANNEL);

    this.sub.on("message", (channel: string, message: string) => {
      if (channel !== ORDER_CHANNEL) return;

      let event: OrderStatusEvent;
      try {
        event = JSON.parse(message) as OrderStatusEvent;
      } catch {
        // Malformed message — ignore silently to avoid crashing the process
        return;
      }

      this.sseManager.sendEvent(event.orderId, "status", event);
    });
  }

  /**
   * Publish an order status change to all subscribed API instances.
   */
  async publishOrderStatus(event: OrderStatusEvent): Promise<void> {
    await this.pub.publish(ORDER_CHANNEL, JSON.stringify(event));
  }

  /**
   * Unsubscribe from the order status channel and clean up listeners.
   */
  async stop(): Promise<void> {
    await this.sub.unsubscribe(ORDER_CHANNEL);
    this.sub.removeAllListeners("message");
  }
}

File: backend/app/realtime/redis_pubsub.py
"""Redis pub/sub for order status changes."""

import json
import logging

from app.redis import get_redis

logger = logging.getLogger(__name__)

CHANNEL_PREFIX = "order_status:"


async def publish_order_status(order_id: str, status: str) -> None:
    """Publish an order status change to Redis."""
    redis = await get_redis()
    if redis is None:
        return
    channel = f"{CHANNEL_PREFIX}{order_id}"
    message = json.dumps({"order_id": order_id, "status": status})
    await redis.publish(channel, message)


async def subscribe_order_status(order_id: str):  # type: ignore[no-untyped-def]
    """Subscribe to order status changes via Redis pub/sub.

    Returns an async generator of status messages.
    """
    redis = await get_redis()
    if redis is None:
        return
    channel = f"{CHANNEL_PREFIX}{order_id}"
    pubsub = redis.pubsub()
    await pubsub.subscribe(channel)
    try:
        async for message in pubsub.listen():
            if message["type"] == "message":
                data = json.loads(message["data"])
                yield data
    finally:
        await pubsub.unsubscribe(channel)
        await pubsub.close()
